lib/publiccloud/basetest.pm

```diff
@@ -103,36 +103,56 @@ sub provider_factory {
 
 sub cleanup {
     # to be overridden by tests
+    # non-destructive cleanup for test module started by post_run_hook
+    # primary for log collection | restore clean state after test module
+    # can and be called after each module run
     return 1;
 }
 
-sub _cleanup {
+sub destroy {
+    # to be overridden by tests
+    # destructive routine in post_fail_hook | last module of test
+    return 1;
+}
+
+sub _destroy {
     my ($self) = @_;
-    die("Cleanup called twice!") if ($self->{cleanup_called});
-    $self->{cleanup_called} = 1;
+    # run module overriden destroy
+    eval { $self->destroy(); } or bmwqemu::fctwarn("self::destroy() failed -- $@");
```

Comment:
I just briefly checked implementation of `destroy` in `azure` . I see that in this implementation there is also call to `SUPER::destroy()` which make it to be called two times ( first here , because `$self->destroy();` is calling parent implementation , and second time below in line 124 )